src/services/index.ts

```diff
@@ -0,0 +1,448 @@
+    if (dbTx) {
+      logger.debug(`Transaction ${hash} already on the database, ignoring..`);
+      mysql.destroy();
+
+      return;
+    }
+
+    let height: number | null = metadata.height;
+
+    if (!isBlock(version) && !metadata.first_block) {
+      height = null;
+    }
+
+    const txOutputs: TxOutputWithIndex[] = prepareOutputs(outputs, tokens);
+    const txInputs: TxInput[] = prepareInputs(inputs, tokens);
+
+    let heightlock = null;
+    if (isBlock(version)) {
+      if (typeof height !== 'number' && !height) {
+        throw new Error('Block with no height set in metadata.');
+      }
+      // unlock older blocks
+      const utxos = await getUtxosLockedAtHeight(mysql, now, height);
+
+      if (utxos.length > 0) {
+        logger.debug(`Block transaction, unlocking ${utxos.length} locked utxos at height ${height}`);
+        await unlockUtxos(mysql, utxos, false);
+      }
+
+      // set heightlock
+      heightlock = height + blockRewardLock;
+
+      // get the first output address
+      const blockRewardOutput = outputs[0];
+
+      // add miner to the miners table
+      // @ts-ignore
+      await addMiner(mysql, blockRewardOutput.decoded.address, hash);
+
+      // here we check if we have any utxos on our database that is locked but
+      // has its timelock < now
+      //
+      // we've decided to do this here considering that it is acceptable to have
+      // a delay between the actual timelock expiration time and the next block
+      // (that will unlock it). This delay is only perceived on the wallet as the
+      // sync mechanism will unlock the timelocked utxos as soon as they are seen
+      // on a received transaction.
+      await unlockTimelockedUtxos(mysql, now);
+    }
+
+    if (version === hathorLib.constants.CREATE_TOKEN_TX_VERSION
+       && token_name
+       && token_symbol) {
+      if (!token_name || !token_symbol) {
+        logger.error('Processed a token creation event but it did not come with token name and symbol');
+        process.exit(1);
+      }
+      await storeTokenInformation(mysql, hash, token_name, token_symbol);
+    }
+
+    // check if any of the inputs are still marked as locked and update tables accordingly.
+    // See remarks on getLockedUtxoFromInputs for more explanation. It's important to perform this
+    // before updating the balances
+    const lockedInputs = await getLockedUtxoFromInputs(mysql, inputs);
+    await unlockUtxos(mysql, lockedInputs, true);
+
+    // add transaction outputs to the tx_outputs table
+    markLockedOutputs(txOutputs, now, heightlock !== null);
+
+    // Add the transaction
+    logger.debug('Will add the tx with height', height);
+    await addOrUpdateTx(
+      mysql,
+      hash,
+      height,
+      timestamp,
+      version,
+      weight,
+    );
+
+    // Add utxos
+    await addUtxos(mysql, hash, txOutputs, heightlock);
+    await updateTxOutputSpentBy(mysql, txInputs, hash);
+
+    // Handle genesis parent txs:
```

Comment:
Why genesis?

Reply:
The comment is misleading, this comment is explaining the if case, genesis tx does not have inputs nor outputs

Refactored the comment in 3a8f4ee660eb344ea56439c0b5270c128d48b950